src/java/org/apache/cassandra/index/sai/plan/StorageAttachedIndexSearcher.java

```diff
@@ -492,56 +524,111 @@ public UnfilteredRowIterator computeNext()
+            // We store a mapping of primary key to the collection of PrimaryKeyWithSortKey objects that will be used
+            // to validate the row.
+            var keysToMaterialize = new HashMap<PrimaryKey, List<PrimaryKeyWithSortKey>>();
+            while (keysToMaterialize.size() < count)
             {
                 var primaryKeyWithSortKey = nextSelectedKeyInRange();
+                // The iterator is exhausted.
                 if (primaryKeyWithSortKey == null)
-                    return;
+                    break;
+                // If we've already processed the key, we can skip it. Because the score ordered iterator does not
+                // deduplicate rows, we could see dupes if a row is in the ordering index multiple times. This happens
+                // in the case of dupes and of overwrites.
+                if (processedKeys.containsKey(primaryKeyWithSortKey.primaryKey()))
+                    continue;
                 var nextPrimaryKey = primaryKeyWithSortKey.primaryKey();
-                var accumulator = keys.computeIfAbsent(nextPrimaryKey, k -> new ArrayList<>());
-                if (primaryKeyPriorityQueue != null && accumulator.isEmpty())
-                    primaryKeyPriorityQueue.add(nextPrimaryKey);
-                accumulator.add(primaryKeyWithSortKey);
+                keysToMaterialize.computeIfAbsent(nextPrimaryKey, k -> new ArrayList<>()).add(primaryKeyWithSortKey);
             }
+
+            if (keysToMaterialize.isEmpty())
+                return true;
+
+            List<CompletableFuture<UnfilteredRowIterator>> rowFutures = new ArrayList<>(keysToMaterialize.size());
+
+            // Iterate through the keys and materialize the partitions using the configured executor.
+            for (var entry : keysToMaterialize.entrySet())
+            {
+                var primaryKey = entry.getKey();
+                var primaryKeyValidators = entry.getValue();
+                CompletableFuture<UnfilteredRowIterator> future = new CompletableFuture<>();
+                rowFutures.add(future);
+                PARALLEL_EXECUTOR.maybeExecuteImmediately(() -> {
+                    // We cancel the future if the query has timed out, so it is worth a quick read to the
+                    // volatile field to possibly avoid getting the partition unnecessarily.
+                    if (future.isDone())
+                        return;
+                    try (var partition = controller.getPartition(primaryKey, view, executionController))
+                    {
+                        // Validates that the parition's row is valid for the query predicates. It returns null
+                        // if the row is not valid. Reasons for invalidity include:
+                        // 1. The row is a range or row tombstone or is expired.
+                        // 2. The row does not satisfy the query predicates
+                        // 3. The row does not satisfy the ORDER BY clause
```

Comment:
How can a row not satisfy the `ORDER BY` clause?

Reply:
I clarify it here and in the comment. Essentially, because of updates, we have to check that the source sstable for the materialized cell is the same sstable as the source SAI index's sstable. Otherwise, we can have out of order rows due to updates to a worse position in the iterator of ordered rows.